### ISSUE TITLE
Looks like complex tensor is broken right now

### ISSUE THREAD

**Issue by williamFalcon** [Roger-luo/pytorch-complex#8]
Hi Guys,

Love the effort with this! I'm happy to help drive it home. I am, however, not familiar with cuda and my cpp is at least 5 years rusty haha. However, with a bit of guidance, I can probably make something happen here.

Like you guys, I'm also a fulltime student, so my contributions might be limited.

I finally compiled the extension but it looks to be broken atm (i suspect it might be a pt issue).
```
>>> from torch_complex import torch
>>>
>>> torch.ones(4,3, dtype=torch.complex64)
tensor([[(5.263544247e-315+5.263544247e-315j), (5.263544247e-315+5.263544247e-315j), (5.263544247e-315+5.263544247e-315j)],
        [(5.263544247e-315+5.263544247e-315j), (5.263544247e-315+5.263544247e-315j), (5.263544247e-315+5.263544247e-315j)],
        [(5.263544247e-315+5.263544247e-315j), (5.263544247e-315+5.263544247e-315j), (5.263544247e-315+5.263544247e-315j)],
        [(5.263544247e-315+5.263544247e-315j), (5.263544247e-315+5.263544247e-315j), (5.263544247e-315-1.4916681462400413e-154j)]],
       dtype=torch.complex64)
>>> torch.ones(4,3, dtype=torch.complex64) * 5
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: empty_strided is not implemented for type CPUComplexTensor<float>
>>>
```

I have implemented all the complex operations in this repo ([/williamFalcon/pytorch-complex-tensor](https://github.com/williamFalcon/pytorch-complex-tensor).   The problem with my implementation is that it just manipulates real tensors as complex and gradients don't match the tensorflow gradient computation...
https://github.com/williamFalcon/pytorch-complex-tensor/blob/master/Complex%20demo%20example.ipynb

I suspect to get this to actually work, I'm going to need to have complex scalar support which seems can only be done in cpp...?  

**The way I see it, we could probably make this happen in 1 of 2 ways:**
1. You guys take in what I have into your code to get the operations working correctly? (this might already be done, but looks broken).
2. Happy to do a crash course in what I need to know to get this implemented into your library (which files to edit, etc...) and I can do the port.
3. ?

I have probably 1-2 days I can allocate to get this done.

**UPDATE** - I was able to figure out how to get gradient support for my project. Still happy to join forces to get PT complex support added.

**Comment by Roger-luo** [comment]
What we are planning to implement here is a cpp based implementation with BLAS/CUBLAS support. And it's in very early stage that you won't be able to use it in practical projects...I compared my current implementation with Julia's broadcast on complex numbers last time, it is about 5x slower.

IMHO, I personally won't take in what you have in https://github.com/williamFalcon/pytorch-complex-tensor/blob/master/complex_tensor.py

I believe there has been a lot similar efforts in Python before this package, I personally don't want to make duplicated effort (e.g https://github.com/PIQuIL/QuCumber/blob/master/qucumber/utils/cplx.py
). You can never have something fast and elegant in pure python...  Python implementation can be much slower than cpp, even without SIMD support.

Thus, I believe these python implementation should stay in where it is for temporary use, so people's research code won't be effected by the instability of this repo (it gonna change a lot... I believe)

https://github.com/PIQuIL/QuCumber/issues/53

You need to compile a master version with this patch (it should work on CPU) working https://github.com/Roger-luo/pytorch-complex/issues/4 to have scalar operation. And please read the development plan back in issue https://github.com/pytorch/pytorch/issues/755 (@ezyang had listed several features we are going to have).

Unfortunately....nothing is guaranteed to work here for now. Last time when I was working on this, I only get some scalar operator and printing working. It is easy to have functions like `sin`/`cos`/etc. working if you know how backend bindings work (it's mainly about generating bindings). I don't have much notes on this however, you could try to read the cpp backend source directly if you can... and ask us in the issue if you meet any cpp problem.

But I'm going to focus on that broadcast and promotion issue next time when I have time on this so we could have things like scalar operations and function work in general.

**Comment by williamFalcon** [comment]
makes sense. I'm just looking through this (https://github.com/PIQuIL/QuCumber/blob/master/qucumber/utils/cplx.py) and it looks very similar to my representation (you use [x,y], I use [[x],[y]]). Both are different ways of doing the same thing which is great.

I'm still concerned that in either approach gradients won't work out. Have you actually tried this in a learning setting? I just compared against tf gradients and didn't get the same result.

**Comment by Roger-luo** [comment]
That's strange, which operator is not correct? Most complex operator should just work as a tuple (or another dimension):

https://github.com/FluxML/Zygote.jl/issues/29

BTW, you could check the gradient by checking its numerical jaccobian with `torch.gradcheck`.

**Comment by williamFalcon** [comment]
![image](https://user-images.githubusercontent.com/3640001/52371677-30f40b00-2a24-11e9-8309-7bf9c8d00673.png)

**Comment by williamFalcon** [comment]
I just did a small test here (https://github.com/williamFalcon/pytorch-complex-tensor/blob/master/Complex%20demo%20example.ipynb) against your cplx library. It didn't match the tf output. (look at last 2 sections).

However, I figured out how to do it with mine. I matched against numpy and tensorflow and looks like we get the same outputs. I'll stick to mine for my research, but happy to chat again about joining forces to give PT complex support.

Thanks for getting started on this!
